src/PrestaShopBundle/Security/Annotation/AdminSecurity.php

```diff
@@ -26,109 +26,178 @@
+     */
+    protected string|Expression $attribute;
+
+    /**
+     * If set, will throw HttpKernel's HttpException with the given $statusCode.
+     * If null, Security\Core's AccessDeniedException will be used.
      */
-    protected $domain = 'Admin.Notifications.Error';
+    protected ?int $statusCode = null;
+
+    /**
+     * If set, will add the exception code to thrown exception.
+     */
+    protected ?int $exceptionCode = null;
 
     /**
      * The route for the redirection.
      *
      * @todo: Once the onboarding page is migrated, set default to his route name.
-     *
-     * @var string|null
      */
-    protected $redirectRoute;
+    protected ?string $redirectRoute = null;
+
+    // The translation domain for the message.
+    protected string $domain = 'Admin.Notifications.Error';
 
     /**
      * @deprecated once the back office is migrated, rely only on route.
      * The url for the redirection
-     *
-     * @return string
      */
-    protected $url = 'admin_domain';
+    protected string $url = 'admin_domain';
 
     /**
-     * The route params which are used together to generate the redirect route.
-     *
-     * @var array
+     * The message of the exception - has a nice default if not set.
      */
-    protected $redirectQueryParamsToKeep = [];
+    protected string $message = 'Access Denied.';
+
+    // The route params which are used together to generate the redirect route.
+    protected array $redirectQueryParamsToKeep = [];
 
     /**
-     * @return string
+     * @param array|string $data
      */
-    public function getDomain()
+    public function __construct(
+        $data = [],
+        string $message = null,
+        string $domain = null,
+        string $url = null,
+        array $redirectQueryParamsToKeep = null,
+        ?int $statusCode = null,
+        ?int $exceptionCode = null,
+        ?string $redirectRoute = null,
+    ) {
+        $values = [];
+        if (is_string($data)) {
+            $values['attribute'] = $data;
+        } else {
+            $values = $data;
+        }
+
+        $values['message'] = $values['message'] ?? $message ?? $this->message;
+        $values['domain'] = $values['domain'] ?? $domain ?? $this->domain;
+        $values['url'] = $values['url'] ?? $url ?? $this->url;
+        $values['redirectQueryParamsToKeep'] = $values['redirectQueryParamsToKeep'] ?? $redirectQueryParamsToKeep ?? $this->redirectQueryParamsToKeep;
+
+        $values['statusCode'] = $values['statusCode'] ?? $statusCode;
+        $values['exceptionCode'] = $values['exceptionCode'] ?? $exceptionCode;
+        $values['redirectRoute'] = $values['redirectRoute'] ?? $redirectRoute;
+
+        foreach ($values as $k => $v) {
+            if (!method_exists($this, $name = 'set' . $k)) {
+                throw new RuntimeException(sprintf('Unknown key "%s" for annotation "@%s".', $k, static::class));
+            }
+
+            $this->$name($v);
+        }
+    }
```

Comment:
We are forced to have a constructor like this, to manage both attributes and annotations, where the parameters are in an array